dll/win32/wdmaud.drv/legacy.c

```diff
@@ -1,89 +1,306 @@
 /*
- * PROJECT:     ReactOS Sound System
- * LICENSE:     GPL - See COPYING in the top level directory
- * FILE:        dll/win32/wdmaud.drv/wdmaud.c
- *
- * PURPOSE:     WDM Audio Driver (User-mode part)
- * PROGRAMMERS: Andrew Greenwood (silverblade@reactos.org)
-                Johannes Anderwald
- *
- * NOTES:       Looking for wodMessage & co? You won't find them here. Try
- *              the MME Buddy library, which is where these routines are
- *              actually implemented.
- *
+ * PROJECT:     ReactOS Audio Subsystem
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     WDM Audio Driver Mapper (User-mode part)
+ * COPYRIGHT:   Copyright 2007-2008 Andrew Greenwood (silverblade@reactos.org)
+ *              Copyright 2009-2010 Johannes Anderwald
+ *              Copyright 2022 Oleg Dubinskiy (oleg.dubinskij30@gmail.com)
  */
 
 #include "wdmaud.h"
 
-#define NDEBUG
-#include <debug.h>
-#include <mmebuddy_debug.h>
+#include <mmixer.h>
 
-#define KERNEL_DEVICE_NAME      L"\\\\.\\wdmaud"
+#define YDEBUG
+#include <debug.h>
 
+extern BOOL CriticalSectionInitialized;
+extern CRITICAL_SECTION WdmAudCriticalSection;
+extern MIXER_CONTEXT MixerContext;
 HANDLE KernelHandle = INVALID_HANDLE_VALUE;
 DWORD OpenCount = 0;
 
+/*
+    This is a wrapper around DeviceIoControl which provides control over
+    instantiated sound devices. It waits for I/O to complete (since an
+    instantiated sound device is opened _In_ overlapped mode, this is necessary).
+*/
+MMRESULT
+WdmAudIoControl(
+    _In_ PWDMAUD_DEVICE_INFO DeviceInfo,
+    _In_opt_ ULONG BufferSize,
+    _In_opt_ PVOID Buffer,
+    _In_ DWORD IoControlCode)
+{
+    MMRESULT Result = MMSYSERR_NOERROR;
+    OVERLAPPED Overlapped;
+    DWORD Transferred = 0;
+    BOOL IoResult;
+
+    /* Overlapped I/O is done here - this is used for waiting for completion */
+    ZeroMemory(&Overlapped, sizeof(OVERLAPPED));
+    Overlapped.hEvent = CreateEvent(NULL, FALSE, FALSE, NULL);
+
+    if ( ! Overlapped.hEvent )
+        return Win32ErrorToMmResult(GetLastError());
+
+    if (DeviceInfo->DeviceType != AUX_DEVICE_TYPE &&
+        DeviceInfo->DeviceType != MIXER_DEVICE_TYPE)
+    {
+        /* Enter critical section */
+        EnterCriticalSection(&WdmAudCriticalSection);
+    }
+
+    /* Store input data */
+    DeviceInfo->BufferSize = BufferSize;
+    DeviceInfo->Buffer = Buffer;
+
+    /* Talk to the device */
+    IoResult = DeviceIoControl(KernelHandle,
+                               IoControlCode,
+                               DeviceInfo,
+                               sizeof(WDMAUD_DEVICE_INFO),
+                               DeviceInfo,
+                               sizeof(WDMAUD_DEVICE_INFO),
+                               &Transferred,
+                               &Overlapped);
+
+    /* If failure occurs, make sure it's not just due to the overlapped I/O */
+    if ( ! IoResult )
+    {
+        if (GetLastError() == ERROR_IO_PENDING)
```

Comment:
`GetOverlappedResult` might be more appropriate than hand-rolling this? In particular, the `Transferred` variable will not currently be set correctly in the async case.

Reply:
This call existed from the start in my code, but I've removed it, because it creates an endless wait with deadlocks, so initialization hangs. Moreover, Windows does not do it. I don't see any `GetOverlappedResult` call in an unassemble of `wdmaudIoControl`. However, that routine does the wait only if `bWait` is `TRUE`. If I'll set it to 'FALSE`, then waiting will not be performed, and the problem will be gone, **but does calling it make any sense then?** :thinking:

Reply:
Regarding `Transferred` variable, your guessing is correct, but if it causes such problems, then maybe I can to not use it at all, because it is used only for a debug print to view the passed buffer size... :thinking: